### PR TITLE
ERC4626 extensions and override for `(TokenRouterComponent)._transfer_from`

### PR DESCRIPTION
**Pull request by EgeCaner**

### CODE REVIEW

File: cairo/src/contracts/token/components/hyp_erc20_collateral_component.cairo [hunk 1]
```diff
@@ -80,6 +81,11 @@ pub mod HypErc20CollateralComponent {
         fn balance_of(self: @ComponentState<TContractState>, account: ContractAddress) -> u256 {
             self.wrapped_token.read().balance_of(account)
         }
+
+        fn get_wrapped_token(self: @ComponentState<TContractState>) -> ContractAddress {
+            let wrapped_token: ERC20ABIDispatcher = self.wrapped_token.read();
```
Comment (ametel01) [inline cairo/src/contracts/token/components/hyp_erc20_collateral_component.cairo:86]: Just a matter of form and to reduce verbosity:

```rust
fn get_wrapped_token(self: @ComponentState<TContractState>) -> ContractAddress {
            self.wrapped_token.read().contract_address
}
```

File: cairo/src/contracts/token/components/hyp_erc721_collateral_component.cairo [hunk 2]
```diff
@@ -39,5 +40,10 @@ pub mod HypErc721CollateralComponent {
         fn balance_of(self: @ComponentState<TContractState>, account: ContractAddress) -> u256 {
             self.wrapped_token.read().balance_of(account)
         }
+
+        fn get_wrapped_token(self: @ComponentState<TContractState>) -> ContractAddress {
```
Comment (ametel01) [inline cairo/src/contracts/token/components/hyp_erc721_collateral_component.cairo:44]: reduce verbosity as above

File: cairo/src/contracts/token/extensions/hyp_erc20_vault.cairo [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+        // Overrides ERC20.balance_of()
+        fn balance_of(self: @ContractState, account: ContractAddress) -> u256 {
+            let balance = ERC20MixinImpl::balance_of(self, account);
+            self.shares_to_assets(balance)
+        }
+
+        fn allowance(
+            self: @ContractState, owner: ContractAddress, spender: ContractAddress
+        ) -> u256 {
+            ERC20MixinImpl::allowance(self, owner, spender)
```
Comment (ametel01) [inline cairo/src/contracts/token/extensions/hyp_erc20_vault.cairo:274]: does `self.erc20.allowance()` not work?

**Review by ametel01 — CHANGES_REQUESTED** [review]
Good work! Address the comment and we can merge

File: cairo/src/contracts/token/extensions/hyp_erc20_vault_collateral.cairo [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        ) -> u256 {
+            let mut contract_state = TokenRouterComponent::HasComponent::get_contract_mut(ref self);
+            TokenRouterHooksTraitImpl::transfer_from_sender_hook(ref self, amount_or_id);
+            let shares = contract_state._deposit_into_vault(amount_or_id);
+            let vault = contract_state.vault.read();
+            let exchange_rate = math::mul_div(
+                PRECISION, vault.total_assets(), vault.total_supply(),
+            );
+            let token_metadata: Bytes =
+                BytesTrait::new_empty(); // TODO: exchange_rate // abi.encode ? should it be big endian?
```
Comment (EgeCaner) [inline cairo/src/contracts/token/extensions/hyp_erc20_vault_collateral.cairo:149]: @ametel01  what do you think about this? solidity does abi.encode(exchange_rate) [here](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/518a1bef91a025c5a2c580134aa5fdb48fdd5667/solidity/contracts/token/extensions/HypERC4626Collateral.sol#L55)

File: cairo/src/contracts/token/components/hyp_erc20_collateral_component.cairo (hunk 1 above)
Comment (EgeCaner) [reply cairo/src/contracts/token/components/hyp_erc20_collateral_component.cairo:86]: Compiler is failing to infer type

**Pull request merged**